hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/ExpressionMath.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.math.MathContext;
+import java.math.RoundingMode;
+
+import static com.hazelcast.sql.impl.type.QueryDataType.MAX_DECIMAL_PRECISION;
+
+/**
+ * Utility methods for math functions.
+ */
+public final class ExpressionMath {
+
+    /**
+     * Math context used by expressions while doing math on BigDecimal values.
+     * <p>
+     * The context uses {@link RoundingMode#HALF_UP HALF_UP} rounding mode, with
+     * which most users should be familiar from school, and limits the precision
+     * to {@link QueryDataType#MAX_DECIMAL_PRECISION}.
+     */
+    public static final MathContext DECIMAL_MATH_CONTEXT = new MathContext(MAX_DECIMAL_PRECISION, RoundingMode.HALF_UP);
+
+    private ExpressionMath() {
+        // No-op.
+    }
+
+    /**
+     * Divides the left-hand side operand by the right-hand side operand.
+     * <p>
+     * Unlike the regular Java division operator, throws an exception if division
+     * resulted in overflow.
+     *
+     * @param left  the left-hand side operand.
+     * @param right the right-hand side operand.
+     * @return a division result.
+     * @throws QueryException if overflow or division by zero is detected.
+     */
+    public static long divideExact(long left, long right) {
+        if (left == Long.MIN_VALUE && right == -1) {
+            throw QueryException.error(SqlErrorCode.DATA_EXCEPTION,
+                    "BIGINT overflow in '/' operator (consider adding explicit CAST to DECIMAL)");
+        }
+        try {
+            return left / right;
+        } catch (ArithmeticException e) {
+            throw QueryException.error(SqlErrorCode.DATA_EXCEPTION, "division by zero");
+        }
+    }
+
+    /**
+     * Divides the left-hand side operand by the right-hand side operand.
+     * <p>
+     * Unlike the regular Java division operator, throws an exception if division
+     * by zero is detected.
+     *
+     * @param left  the left-hand side operand.
+     * @param right the right-hand side operand.
+     * @return a division result.
+     * @throws QueryException if division by zero is detected.
+     */
+    public static double divideExact(double left, double right) {
+        double result = left / right;
+        if (Double.isInfinite(result)) {
+            throw QueryException.error(SqlErrorCode.DATA_EXCEPTION, "division by zero");
+        }
+        return result;
```

Comment:
We can return NaN if both operands are 0, is that ok?

Reply:
fixed 084c08a2d61de582664c629d7412d31dac006e86